### ISSUE TITLE
Run cleaner concurrently also inside one archive 

### ISSUE THREAD

**Issue by pmoravec** [sosreport/sos#3097]
`sos report --cleaner` cleans the generated report sequentially using one thread, what prolongs its already "naturally slow" execution. We should run it concurrently, similarly like `sos report` executes plugins.

I see two approaches here:
- run individual parsers (IP, MAC, username,..) concurrently, with a lock on each file they update. This is required to prevent concurrent modification of the same file by multiple parsers.
- or for each parser running sequentially, clean files in parallel

Currently, `cleaner` iterates over all files and runs individual parsers sequentially for one file after another. That is quite away of either above approach.

Also, `sos clean --jobs=4` declares concurrency on individual archives that can be cleaned in parallel. This additional concurrency (with assumed 4 workers as default) should be aligned with that one - we should not end up with 4*4 concurrent workers running in parallel.

Despite there are the above obstacles to the proposal, I think the justification for the improvement is solid: the use cases when `cleaner` is run on a sequential archive dominate and the improvement would speed them up significantly.

**Comment by TurboTurtle** [comment]
I took a stab at moving our concurrency to process-based a while back and the direction I took there was to try and assign each archive to a specific process. There were a few challenges there - in no particular order:

- We need to keep the parsers sync'd across processes, so that obfuscations are kept consistent across both archives and files.
- Because of this, we need to use shared memory objects for holding those obfuscations. For multiple archives this is fine, but in the case of single archive cleaning it actually significantly harms performance. 
- Logging is performed via the archive abstractions - but logging streams are not able to be serialized and so the pickling that occurs when forking out to a new process fails. There was/is some ambiguity here as newer versions of python don't seem to have the same problem (the new process initializes without issue) but core python maintainers have stated that _shouldn't_ be happening.

That being said, I'm not of the opinion that we have to have parallelism at the archive level. If we instead move to file-level parallelism, that's fine - we just need to be able to report when we've finished the last file in an archive and when we've started a new archive's file list.

This also doesn't have to be with a specific library, but it _does_ have to be available for python-3.6 for our downstreams that use that version.

**Comment by NikhilKakade-1** [comment]
@TurboTurtle @pmoravec , any new developments to fix it?

**Comment by pmoravec** [comment]
> @TurboTurtle @pmoravec , any new developments to fix it?

Nothing I am aware of..

**Comment by NikhilKakade-1** [comment]
@TurboTurtle  @pmoravec , any update?
Is [3262](https://github.com/sosreport/sos/pull/3262) some how related to it ?

**Comment by TurboTurtle** [comment]
@NikhilKakade-1 the short update is - no update.

I've since left RH and my time to dedicate to sos is as such much reduced. Before I left however, there was some discussion on moving to a queue-based systems for obfuscations. The basic idea is that we'd have multiple processes (as opposed to threads) spawned which trawl through an archive/file looking for matches via the parsers. On a match, an item is added to a queue running in the main thread which handles the obfuscation, and then returns the obfuscated match to the process waiting for it.

While mostly straight forward in theory, this is a major change to implement and test. I am not sure when (or if) I'd have the time available to really dedicate to this effort myself in the near future. We (as in the sos project) would likely need resource(s) from our commercial sponsors (namely RH/Canonical) to be able to get this kind of work hammered out in a reasonable timeframe.

**Comment by NikhilKakade-1** [comment]
OK, Thanks !!

@pmoravec, @arif-ali  anything from your side on the above comment?

**Comment by pmoravec** [comment]
> OK, Thanks !!
> 
> @pmoravec, @arif-ali anything from your side on the above comment?

No promise here. We *might* have some more resource for `sos` development in general, but esp. for such major change, I dont want to raise expectations we dont need to meet.

**Comment by arif-ali** [comment]
> OK, Thanks !!
> 
> @pmoravec, @arif-ali anything from your side on the above comment?

I would love to see better parallelism and concurrency, but I don't have bandwidth to take up on anything like that. But, as it has been raised this would require significant change

**Comment by ijajmulani** [comment]
any update on this issue. I'm also facing is issue.

**Comment by pmoravec** [comment]
> any update on this issue. I'm also facing is issue.

Apart of some sporadic discussion at https://github.com/orgs/sosreport/discussions/3476 , no update. No final consensus on the way of implementation (Jake's original idea is followed by two other ones), no work force for it.

**Comment by NikhilKakade-1** [comment]
@TurboTurtle / @pmoravec , To address this, I've been contemplating the implementation of a wrapper script.

The script starts by retrieving a list of enabled plugins using the `sos report -l` command. It then proceeds to execute each plugin individually, utilizing the `-o` flag along with the `--clean` option, allowing for concurrent execution. This approach facilitates direct cleaning of the files collected by each plugin, potentially streamlining the data collection and obfuscation process. With this setup, there's no need to wait for the completion of each plugin before initiating the cleaning process. Additionally, dividing the process into multiple plugins enables parallelism on a batch basis, further enhancing efficiency. Afterward, the resulting multiple tarballs can be consolidated into a single tarball.
However, progress is currently hindered by an issue identified as #3433.

I'm looking forward to receiving your inputs and suggestions

Thanks

**Comment by TurboTurtle** [comment]
That wouldn't be a design that gets picked up by us. To many moving parts, and the reassembly of potentially dozens of archives will, by nature, always be fragile.

Also, as you note, the obfuscation of an archive is based on the contents of the archive. If we want to obfuscate hostnames, we need to know the hostname of the system so we can identify other hostnames we would also want to obfuscate based on the FQDN - and that requires either the information to be present in the archive, to be passed by the user at runtime, or for that information to already be recorded in an obfuscation map in `/etc/sos/cleaner/default` (or similar saved mapping location). This doesn't lend itself well to wrappers that attempt to fragment the workflow.

I still believe that the best path forward is what I describe, admittedly at a high level, in [the discussion on cleaner performance](https://github.com/orgs/sosreport/discussions/3476#discussioncomment-8225987) - we should leverage the native concurrency features of python across multiple processes instead of threads.

To be blunt, this work needs to be focused on by someone who is very familiar with process-based concurrency in python and who is able to commit full-time development to it for a decent stretch of time. I myself have not been able to do so since leaving Red Hat.

**Comment by NikhilKakade-1** [comment]
I think we can have the require file/cmd output with --clean, 
https://github.com/sosreport/sos/issues/3433#issuecomment-2048305532

**Comment by pmoravec** [comment]
> I still believe that the best path forward is what I describe, admittedly at a high level, in [the discussion on cleaner performance](https://github.com/orgs/sosreport/discussions/3476#discussioncomment-8225987) - we should leverage the native concurrency features of python across multiple processes instead of threads.
> 
> To be blunt, this work needs to be focused on by someone who is very familiar with process-based concurrency in python and who is able to commit full-time development to it for a decent stretch of time. I myself have not been able to do so since leaving Red Hat.

What about the [alternative approach](https://github.com/orgs/sosreport/discussions/3476#discussioncomment-8501581) I suggested in the discussion? The processes-based concurrency approach might be most elegant solution but we strive finding a volunteer for it for approx one year, unsuccessfully. Implementing my approach should be much easier (please correct me if my estimation is wrong) and final solution can be still acceptable as a long term solution. Well, with two potential gotchas:

1) some UX regression with moving from obfuscated strings like `host7` to strings like `host-4B9F4B0B` that is less human readable
2) Using salt in clustered environment

**Comment by NikhilKakade-1** [comment]
https://github.com/sosreport/sos/blob/main/sos/cleaner/parsers/__init__.py#L97

In my opinion, optimizing the data masking process is essential. In terms of computational complexity, the worst-case scenario is O(k * n * m), where k is the number of lines, n is the number of compiled regexes in self.mapping.compiled_regexes, and m is the average length of the lines. Conversely, the best case is O(k * m).

I've been mulling over a few questions. 

1. Should we consider periodic cleanup of mapping files as the number of compiled regexes increases over time, potentially impacting total processing time?
2. What factors should be considered when deciding between processing find and replace operations line by line versus in batches?

**Comment by pmoravec** [comment]
The complexity theory approach reminds me my university studies :)

We dont have any cleanup of mapping files. It can have sense, but how/when to decide cleanup should be done? Every month? When mapping is bigger than X records? Remove selective records when they have not been found&replaced in past Y cleaner runs? Each heuristic has its pros and negs. Anyway that is independent topic (that I am definitely not against it) to the limited cleaner concurrency.

Anyway, the number of compiled regexes is usually the much much smallest number in that equation. We might better try preventing cleaners to touch files we know as "safe" (https://github.com/sosreport/sos/blob/main/sos/cleaner/__init__.py#L775-L780), and optionally skip processing files if no parser is applicable. As e.g. can `/var/lib/selinux` contain any sensitive info? Or `/sys/fs/cgroup`, `/proc/sys`  or `/sys/kernel`? I expect some of the dirs can contain some sensitive info, I am not an expert on either, and we should be rather conservative than risk we skip obfuscating anything here, but maybe this is also a room for improvement..? Again, topic independent on the original "make cleaner (more) concurrent".

**Comment by TrevorBenson** [comment]
> I've been mulling over a few questions.
> 
> 1. Should we consider periodic cleanup of mapping files as the number of compiled regexes increases over time, potentially impacting total processing time?

Yes? However, retaining the same obfuscation is very useful in practice. 

Removing selective records not seen in a long time as @pmoravec mentions is one option that could be viable. This presents separate issues, however, like tracking the "temperature" or last-seen timestamp for each obfuscation. This should only be added as a non-default option, allowing the choice of when to enable a cleanup.

## Separately

I like the deterministically computed hash approach @pmoravec discusses in [#3476_comment-8501581](https://github.com/orgs/sosreport/discussions/3476#discussioncomment-8501581), especially if a solution to the problem of using separate salts per host in a single collection is solved. As mentioned, without solving this it presents a problem for a general analysis of separate archives in an sos collection.

The difficulty can also increase exponentially when analyzing cluster configurations, especially for clusters that continue to grow the number of total members. Also, a different salt between separate collections would also make it quite difficult to compare collections from different times against one another, which can also be useful in practice.

**Comment by pmoravec** [comment]
> using separate salts per host in a single collection is solved. 

Assuming salt is somehow stored on each host (and just not shared in sosreport, of course), `sos collect --clean` can read the salt from each host before (or after?) collection of the sosreport archive.

This still does not solve use cases like:
- "generate sosreport on one host, then run cleaner on another", where salt from another host is used to sosreport from the first host
- `sos collect --clean` fetches a sosreport from a host that has not generated salt yet as no cleaner has been run there, so a local (new?) salt is used for that sosreport; later on the host runs `sos clean` that generates a new different salt..? (this has solution: when collector detects a host does not have a salt, it generates one and saves it for the host)

There can be other scenarios I haven't imagined, which makes the idea bit fragile. Thats the cons, compared to the `Queue()` idea described in https://github.com/orgs/sosreport/discussions/3476#discussioncomment-8225987 . Pros is much simplier implementation.

I would agree with either approach (if we are certain it has no gotchas and have somebody who can dedicate time to implement it). I dont need my idea to be implemented, I am very pragmatical - let implement anything we can and agree on.

**Comment by NikhilKakade-1** [comment]
> Anyway, the number of compiled regexes is usually the much much smallest number in that equation. We might better try preventing cleaners to touch files we know as "safe"

Skipping folders like /sys , ... during masking seems reasonable, despite containing a large number of files, their overall size is small. I believe this does not effectively increase performance. The time required for masking is directly proportional to the number of substitutions made and the size of the files, as we do check line by line.

**Comment by NikhilKakade-1** [comment]
> Removing selective records not seen in a long time as @pmoravec mentions is one option that could be viable. This presents separate issues, however, like tracking the "temperature" or last-seen timestamp for each obfuscation. This should only be added as a non-default option, allowing the choice of when to enable a cleanup.

Is there any `Issue` created ? If not can we create one, so that we won't miss it?

**Comment by pmoravec** [comment]
> > Removing selective records not seen in a long time as @pmoravec mentions is one option that could be viable. This presents separate issues, however, like tracking the "temperature" or last-seen timestamp for each obfuscation. This should only be added as a non-default option, allowing the choice of when to enable a cleanup.
> 
> Is there any `Issue` created ? If not can we create one, so that we won't miss it?

This has been implemented already, see `--skip-cleaning-files` option and https://github.com/sosreport/sos/pull/3520 .

**Comment by TrevorBenson** [comment]
> > > Removing selective records not seen in a long time as @pmoravec mentions is one option that could be viable. This presents separate issues, however, like tracking the "temperature" or last-seen timestamp for each obfuscation. This should only be added as a non-default option, allowing the choice of when to enable a cleanup.
> > 
> > 
> > Is there any `Issue` created ? If not can we create one, so that we won't miss it?
> 
> This has been implemented already, see `--skip-cleaning-files` option and #3520 .

I don't believe these are the same topic.

* #3520 is about skipping the cleaning of sensitive information.

Where my statement:

> This should only be added as a non-default option, allowing the choice of when to enable a cleanup.

Was a response to [this comment bullet point 1](https://github.com/sosreport/sos/issues/3097#issuecomment-2076921848)
> 1. Should we consider periodic cleanup of mapping files as the number of compiled regexes increases over time, potentially impacting total processing time?

My statement was intended to mean that I did not believe we should automatically cleanup mapping files (i.e. remove mappings from the file which had not been seen in a long time). I felt that if we started to track the temperature (recent usage of each obfuscation mapping) the removal of a previous mapping  should be left up to the user to execute when they wanted to cleanup mappings.

If we have a consistent mapping as discussed in other threads this is probably not a big deal, as recreation of the obfuscation would result in an identical mapping anyway.

**Comment by NikhilKakade-1** [comment]
Thanks @TrevorBenson , yep I was referring to this part.

**Comment by pmoravec** [comment]
Ah I see, sorry for my confusion. Yes, it is worth raising a separate issue for the periodic cleanup of mapping files.

**Comment by TrevorBenson** [comment]
I went ahead and put together #3715

**Comment by pmoravec** [comment]
There are numerous ideas how to implement this (that I confusingly presented in a different issue):

- Jake's cleaner thread (that does not fix the https://github.com/sosreport/sos/issues/3806)
- Pavel's hashing approach (completely orthogonal to anything else, but with own gotchas, esp. risk of guessing whole mapping that is deterministically computable anywhere (with right salt))
- Nikhil's idea of interactivelly writing the mapping with locks (https://github.com/sosreport/sos/issues/3806#issuecomment-2457345169)
- mapper process instead of Jake's thread where cleaners communicate with it via Unix file sockets (https://github.com/orgs/sosreport/discussions/3476#discussioncomment-11779472) - optionally extendable to listen on TCP socket for distributed cleaners (https://github.com/orgs/sosreport/discussions/3476#discussioncomment-11792375)
- cleaner processes/threads "compete" for the just-next hashed value by moving files into the right filename (https://github.com/orgs/sosreport/discussions/3476#discussioncomment-11779499)
- dont reinvent the wheel and use SQLite DB for the "tell me value or assign new one" transaction (https://github.com/orgs/sosreport/discussions/3476#discussioncomment-11779506)


I explored the latest option a bit and have a PoC - see next comment.

**Comment by pmoravec** [comment]
### Proof of Concept of using `sqlite3`:

This script can run concurrently, and it will assign obfuscated numbers for randomly generated hostnames (in form `realhost_[0-99]_[0-999]`. These numbers can be used for the obfuscated strings like `host007`. The same can be extended to any other mapping (with some specifics for e.g. IP mapper that does not have sequential-only "obfuscated numbers").

```
import sqlite3, random, time

con = sqlite3.connect("cleaner_mapping.db")
con.execute('pragma journal_mode=wal') # to improve DB locking on frequent concurrent writes
cur = con.cursor()
cur.execute("CREATE TABLE IF NOT EXISTS shortnames(counter, hostname)") # we can unify the value from hostname to name, for other mappers
cur.execute("CREATE UNIQUE INDEX IF NOT EXISTS ind_count_uniq ON shortnames (counter)")
cur.execute("CREATE UNIQUE INDEX IF NOT EXISTS ind_hostname_uniq ON shortnames (hostname)")
con.commit()

def load_new_entries_from_db(mapping, counter):
    res = cur.execute(f"SELECT * FROM shortnames WHERE counter>={counter}")
    for cnt, value in res.fetchall():
        mapping[value] = cnt
        print(f"Discovered {value} -> {cnt}")

def set_new_value(mapping, hostname):
    while not mapping.get(hostname, False):
        try:
            counter = len(mapping) + 1 # we should keep the counter in mapping class as it does not need to be sequential (for IP addresses), but for this PoC it is sufficient
            cur.execute(f"INSERT INTO shortnames VALUES ({counter}, '{hostname}')")
            con.commit()
            mapping[hostname] = counter
        except sqlite3.IntegrityError:
            print(f"Raised race error when trying to insert {hostname} -> {counter}")
            load_new_entries_from_db(mapping, counter)
        except sqlite3.OperationalError:
            print(f"Raised db lock error, sleeping 1s..")
            time.sleep(1)
            load_new_entries_from_db(mapping, counter)


mapping = {}
load_new_entries_from_db(mapping, len(mapping) + 1)

for i in range(100):
    for j in range(100):
        hostname = f"realhost_{i}_{random.randint(0, 999)}"
        if hostname not in mapping.keys():
            set_new_value(mapping, hostname)
            new = " (new)"
        else:
            new = ""
        print(f"{hostname} -> {mapping[hostname]} {new}")

print("======")
load_new_entries_from_db(mapping, len(mapping) + 1)
print(mapping)
```

There is a tiny risk of db lock when massive parallel writes happen. This is substantially improved by the WAL. Currently, I am unable to hit a DB lock when running the script 10 times concurrently.

Performance: not great, not terrible. The 10 processes run for 15ish seconds to concurrently generate 10k records each (some of them are same, most of them unique). This performance is vastly improved for subsequent runs, when less new hostnames are discovered as new / not yet in mapping.

**Comment by TrevorBenson** [comment]
> There is a tiny risk of db lock when massive parallel writes happen. This is substantially improved by the WAL. Currently, I am unable to hit a DB lock when running the script 10 times concurrently.

I don't hit any DB lock when testing up to 4 concurrent, but at around 5-6 concurrent runs I have been able to hit a db lock.

```
[ ... ]
5043, 'realhost_99_828': 35044, 'realhost_99_941': 35045, 'realhost_99_493': 35046, 'realhost_99_213': 35047, 'realhost_99_748': 35048, 'realhost_99_190': 35049, 'realhost_99_357': 35050, 'realhost_99_66': 35051, 'realhost_99_353': 35052, 'realhost_99_685': 35053, 'realhost_99_44': 35054}

real	0m16.217s
user	0m0.283s
sys	0m1.038s
Raised db lock error, sleeping 1s..
Raised db lock error, sleeping 1s..
Raised db lock error, sleeping 1s..
Raised db lock error, sleeping 1s..
Raised db lock error, sleeping 1s..
```

I tested on a mobile and desktop CPU for comparison, both with 8 cores 16 threads and Max GHz from 4.4 - 5.1. After 8 passes on each I had 3 runs on one and 4 runs on the other platform which hit the DB lock. 

```
for i in {1..5} ; do sleep 1 ; time python poc.py & done
```

FWIW it appeared as if I could get a one to two more in parallel when the tuned profile was set to latency-performance, instead of balanced or lower. However, this is mostly anecdotal after observing 3 runs in a row without hitting a lock. Due to time constraints I didn't do full benchmark runs w/ 10 per set of variables then drop the best and worst and compare averages.

**Comment by pmoravec** [comment]
So it might not scale well. The main problem is the initial filling the mapping since the DB is in fact used "just" as a cache of new entries. I can try improving it such that prepper initially requests a bulk of new records ("insert into the db all hostnames found in `/etc/hosts` and in ..") - that will decrease the number of concurrent transactions.

Also, the script mimics the worst case where a cleaner requests a new record VERY frequently, all the time. That is usually not happening (apart of the initial phase that I can try to improve).

Anyway, I can try replacing the sqlite by the "try moving file to the right location" approach and compare performance (and FS locks).

**Comment by TrevorBenson** [comment]
> So it might not scale well. The main problem is the initial filling the mapping since the DB is in fact used "just" as a cache of new entries. I can try improving it such that prepper initially requests a bulk of new records ("insert into the db all hostnames found in `/etc/hosts` and in ..") - that will decrease the number of concurrent transactions.
> 
> Also, the script mimics the worst case where a cleaner requests a new record VERY frequently, all the time. That is usually not happening (apart of the initial phase that I can try to improve).
> 
> Anyway, I can try replacing the sqlite by the "try moving file to the right location" approach and compare performance (and FS locks).

I'm not against going with the sqlite solution. I only wanted to provide some testing and feedback for the proof of concept. I also think the POC was more likely to result in concurrent ops and hit a lock.

In the end anything you think is resilient enough I would be happy with. I'll also provide additional testing for any other POC you put together.

**Comment by pmoravec** [comment]
### Proof of Concept of using moving files:

The script is variant of the PoC for `sqlite3`, but uses "who moves the file first, they win" approach:


```
import random
from pathlib import Path
import tempfile
import os

DIR = "/root/CONCURRENT-CLEANER/cleaner-filecache/shortnames"
Path(DIR).mkdir(parents=True, exist_ok=True)

def load_new_entries_from_db(mapping, counter):
    fname = os.path.join(DIR, f"{counter}")
    while os.path.isfile(fname):
        with open(fname, 'r') as f:
            value = f.read()
            mapping[value] = counter
            print(f"Discovered {value} -> {counter}")
        counter += 1
        fname = os.path.join(DIR, f"{counter}")

def set_new_value(mapping, hostname):
    tmpfile = None
    while not mapping.get(hostname, False):
        if not tmpfile:
            tmpfile = tempfile.NamedTemporaryFile()
            with open(tmpfile.name, 'w') as f:
                f.write(hostname)
        try:
            counter = len(mapping) + 1 # we should keep the counter in mapping class as it does not need to be sequential (for IP addresses), but for this PoC it is sufficient
            os.link(tmpfile.name, os.path.join(DIR, f"{counter}"))
            mapping[hostname] = counter
        except FileExistsError:
            print(f"Raised race error when trying to insert {hostname} -> {counter}")
            load_new_entries_from_db(mapping, counter)


mapping = {}
load_new_entries_from_db(mapping, len(mapping) + 1)

for i in range(100):
    for j in range(100):
        hostname = f"realhost_{i}_{random.randint(0, 999)}"
        if hostname not in mapping.keys():
            set_new_value(mapping, hostname)
            new = " (new)"
        else:
            new = ""
        print(f"{hostname} -> {mapping[hostname]} {new}")

print("======")
load_new_entries_from_db(mapping, len(mapping) + 1)
print(mapping)
```

Performance: 10%-ish better than `sqlite3`. No (db or FS) locks when used 8 times concurrently. Potentially, we might hit some inode / FS limitation in a scaled environment when we create 100k-s of files..?

I will run some more tests but this seems promising.

**Comment by pmoravec** [comment]
I wanted to give a try to the daemon-like approach, where a process listens on socket for cleaners clients requests and the process manages and builds the mapping. BUT:

- the process must manage clients connections concurrently
- using multi-threaded approach means a multi-thread synchronization just within the "server process" - possible but too complex
- optionally, the only threat can iterate over the pool of connections and manage them one after another: not sure about scalability and complexity
- optionally, each client's communication can be "connect, send request, receive response, drop connection" to keep max 1 connection at any time - but this would be VERY slow (I run a few tests on PoC of PoC).

So I am giving up an attempt for PoC here. Such approach can fix #3476 but.. it would be too huge hammer to today's nail.


From the list of ideas in https://github.com/sosreport/sos/issues/3097#issuecomment-2582708872:

- Jake's approach doesnt work for concurrent programs
- my idea of hashing: I am too afraid no sufficiently reliable hashing algorithm can exist; also outcome would be human unfriendly obfuscated strings
- Nikhil's idea imho does not scale (in mapping file length and in concurrent processes)
- daemon-like approach is too complex
- Adam's "move file" idea sounds to be the best
- Bryn's sqlite3 wins my 2nd place but suffers sporadic db locks and is less performant than "move file"

So until you disagree, I will start implementing the "move file" approach.

Please comment or upvote for (dis)agreement.

**Comment by pmoravec** [comment]
Hi @arif-ali , @TurboTurtle or @TrevorBenson . I would like to implement this feature, to have cleaner able to run concurrently (both in independent processes on the same system, and also multiple threats within one cleaner execution). The current execution times of cleaner are really limiting in some use cases.

From the above exploration of possible implementations, I would vote for the "move files" approach. Please ACK this choice or suggest other idea or upvote other proposed idea to have some agreement soon.

Thanks in advance.

**Comment by TrevorBenson** [comment]
> Hi [@arif-ali](https://github.com/arif-ali) , [@TurboTurtle](https://github.com/TurboTurtle) or [@TrevorBenson](https://github.com/TrevorBenson) . I would like to implement this feature, to have cleaner able to run concurrently (both in independent processes on the same system, and also multiple threats within one cleaner execution). The current execution times of cleaner are really limiting in some use cases.
> 
> From the above exploration of possible implementations, I would vote for the "move files" approach. Please ACK this choice or suggest other idea or upvote other proposed idea to have some agreement soon.
> 
> Thanks in advance.

👍🏼 ACK. I definitely do not want to delay resolution of this issue due to #3476 discussions.

---

I could see this being used in a distributed cleaning. However, that would add a dependency of fuse-sshfs which I haven't used in over 10 years.

Unless others felt that sshfs and the new dependency would be worthwhile and performant to provide a distributed cleaning w/ shared obfuscation map I wouldn't pick it as a potential solution to #3476, but based on the view of others it may be worth a POC to try it out.

**Comment by jcastill** [comment]
I've been checking both approaches (moving files, sqlite) and I think I _mostly_ managed to get rid of the blocking issues using the following options:

```
con.execute('pragma temp_store = memory')
con.execute('pragma synchronous = normal')
con.execute('pragma mmap_size = 30000000000')
```

From [here](https://phiresky.github.io/blog/2020/sqlite-performance-tuning/). I wrote 'mostly' because in my tests (have been running them for about a week, in Fedora 41, CentOS Stream 10, and RHEL 9) two things happen:
- Running with 4 threads never hit the "Raised db lock error, sleeping 1s..", and
- Running 10+ threads prints that message, but only a couple of times every couple of hundred tests more or less, and it doesn't block - waits 1 second, then continues. 

Most probably I ran the tests wrong or missed some lock errors, so take this with a pinch/bucket of salt. But if anyone else can test, I imagine this could help make the decision to which approach to take.

**Comment by pmoravec** [comment]
This means the SQLite approach is a better option, I guess? I will compare performance to know more.

@arif-ali , @TurboTurtle any preferences or opinion what option to chose? I would like to move forward and start implementing it.

Apart of the concurrency implementation, there is also a question of granularity in the use case "(default) four threads in a single `sos clean` process": how the "each mapper to be applied to each file in archive" should be split among the threads?

I think the best approach is to list all files to clean and split them to individual threads (by number of files or weighted by filesize length). An alternative in "split the work per mappers" can lead to very uneven work load and threads stepping on each other toes (modifying same file concurrently).

**Comment by pmoravec** [comment]
I made some progress here!

I have a draft of code changes based on "move files" approach that shrunk cleaner runtime from `2m11.643s` to `0m36.887s` (on 4 CPUs with default `--jobs=4`).

Cleaner (and component.py and presets and ..) have many traps that block calling simple:

```
with concurrent.futures.ProcessPoolExecutor(max_workers=self.opts.jobs) as executor:
    executor.map(self.obfuscate_file, self.archive.get_file_list())
```

within `SoSCleaner` class. The traps are e.g. overwriting some _* methods that prevents __repr__ or __iter__ or other native Python methods to successfully fork a new process and copy all used classes etc.

After digging into that rabbit hole for some time, I decided to move basically whole `SosCleaner.obfuscate_file` into `SoSObfuscationArchive` class.

@TurboTurtle , (or anybody other willing to chat about class boundaries) is that OK for you?

Since forking also `SoSCleaner` means forking whole `SoSComponent`, options, presets and a lot of stuff. And frankly speaking, that classes are not needed for pure "apply these parsers to this file" task that we require from the child processes.

(I wanted to have also sql based approach but that needs some further change - I must create connection to sql DB *after* the forking, for obvious reasons)
